hack/generator/pkg/codegen/pipeline_convert_allof_and_oneof_to_objects.go

```diff
@@ -0,0 +1,542 @@
+
+	case *astmodel.ResourceType:
+		name := fmt.Sprintf("resource%v", i)
+		propertyName := s.idFactory.CreatePropertyName(name, astmodel.Exported)
+		jsonName := s.idFactory.CreateIdentifier(name, astmodel.NotExported)
+		return astmodel.NewPropertyDefinition(propertyName, jsonName, concreteType).MakeOptional(), nil
+
+	default:
+		return nil, errors.Errorf("unexpected oneOf member, type: %T", from)
+	}
+
+}
+
+func (s synthesizer) intersectTypes(left astmodel.Type, right astmodel.Type) (astmodel.Type, error) {
+	for _, handler := range intersectHandlers {
+		result, err := handler(s, left, right)
+		if err != nil {
+			return nil, err
+		}
+
+		if result != nil {
+			return result, nil
+		}
+	}
+
+	return nil, errors.Errorf("don't know how to intersect types: %s and %s", left, right)
+}
+
+// intersectHandler knows how to do intersection for one case only
+// it is biased to examine the LHS type (but some are symmetric and handle both sides at once)
+type intersectHandler = func(synthesizer, astmodel.Type, astmodel.Type) (astmodel.Type, error)
+
+// handles the same case for the right hand side
+func flip(f intersectHandler) intersectHandler {
+	return func(s synthesizer, left astmodel.Type, right astmodel.Type) (astmodel.Type, error) {
+		return f(s, right, left)
+	}
+}
+
+var intersectHandlers []intersectHandler
+
+func init() {
+	intersectHandlers = []intersectHandler{
+		synthesizer.handleEqualTypes, // equality is symmetric
+		synthesizer.handleAnyType, flip(synthesizer.handleAnyType),
+		synthesizer.handleAllOfType, flip(synthesizer.handleAllOfType),
+		synthesizer.handleTypeName, flip(synthesizer.handleTypeName),
+		synthesizer.handleOneOf, flip(synthesizer.handleOneOf),
+		synthesizer.handleOptionalOptional,                           // symmetric
+		synthesizer.handleOptional, flip(synthesizer.handleOptional), // needs to be before Enum
+		synthesizer.handleResourceType, flip(synthesizer.handleResourceType),
+		synthesizer.handleEnumEnum, // symmetric
+		synthesizer.handleEnum, flip(synthesizer.handleEnum),
+		synthesizer.handleObjectObject, // symmetric
+		synthesizer.handleMapMap,       // symmetric
+		synthesizer.handleMapObject, flip(synthesizer.handleMapObject),
+	}
+}
+
+func (s synthesizer) handleOptional(left astmodel.Type, right astmodel.Type) (astmodel.Type, error) {
+	if leftOptional, ok := left.(*astmodel.OptionalType); ok {
+		// is this wrong? it feels wrong, but needed for {optional{enum}, string}
+		return s.intersectTypes(leftOptional.Element(), right)
+	}
+
+	return nil, nil
+}
+
+func (s synthesizer) handleResourceType(left astmodel.Type, right astmodel.Type) (astmodel.Type, error) {
+	if leftResource, ok := left.(*astmodel.ResourceType); ok {
+		if rightResource, ok := right.(*astmodel.ResourceType); ok {
+			// merge two resources: merge spec/status
+
+			spec, err := s.intersectTypes(leftResource.SpecType(), rightResource.SpecType())
```

Comment:
We do this even if `s.SpecOrStatus == chooseStatus`? I'm a bit thrown because you seem to look at those later in this method but not here?

Reply:
oh I see, there's two cases here: left + right are both resources, or left is a resource and right isn't?

Reply:
Yes, if both are resources we merge them fully, otherwise we need to know if the type that contains this resource is a spec or status type. It’s a bit weird. I think it’s actually always Spec as the Status types don’t embed other resources.

Reply:
I should pull out the 2-resource case as `handleResourceResource`.

Reply:
done